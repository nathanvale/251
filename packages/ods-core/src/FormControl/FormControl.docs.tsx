import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Checkbox } from "../Checkbox/Checkbox";
import { FormControl, FormControlProps } from "./FormControl";

export const docs: ComponentDocs<FormControlProps> = {
  category: "Atomic",
  componentName: "FormControl",
  description: [
    "This is an important atomic component that can be used as a wrapper",
    "to build more complicated for component. Examples of using this component",
    "include wrapping a number of Checkboxes or Radios which need to be grouped.",
    "This component has a context provider that allows its children to check its stats",
    "(such as 'error', 'disabled', ...) and style themselves accordingly.",
    "FormLabel, FormHelperText, Radio and Checkbox automatically respond to",
    "'error' and 'disabled' states of this wrapper component if provided.",
    "Notice that the helperTexts for each Radio and Checkbox are NOT changing",
    "colour to if the wrapping FormControl is in error state. This is deliberate and according to our style-guide",
  ].join("\n"),
  propDescriptions: {
    id: "The unique id of the component, required to accessibility",
    children: [
      "Any elements, however with the exception of Radios and Checkboxes",
      "this component should only have 1 input child",
    ].join(" "),
    component: ["The html element type to be used under the hood."].join(" "),
    disabled: [
      "If true will set the state of its input children to disabled as well.",
      "use this to set the state of a group of checkboxes or Radios.",
    ].join(" "),
    fullWidth: [
      "If true the FormControl will take up the full width of its container. This value is also passed down to",
      "the children of FormControl via FormStateContext.",
    ].join(" "),
    error: [
      "If true will set the state of its input children to error as well.",
      "use this to set the state of a group of checkboxes or Radios.",
    ].join(" "),
  },
  migrationGuide: false,
  examples: {
    default: {
      Code: () => <FormControl>This is a fieldset</FormControl>,
    },
    additional: [
      {
        label: "Custom component",
        description: ["Render a div element"].join(" "),
        Code: () => <FormControl component="div">This is a div</FormControl>,
      },
      {
        label: "Complex use-case, error state",
        description: [
          "Checkboxes under the hood respond to 'error' states of this component",
        ].join(" "),
        Code: () => (
          <FormControl error>
            <Checkbox id="chk1" label="Check 1" helperText="Some description" />
            <Checkbox id="chk2" label="Check 2" />
          </FormControl>
        ),
      },
      {
        label: "Complex use-case, disabled state",
        description: [
          "Checkboxes under the hood respond to 'disabled' states of this component",
        ].join(" "),
        Code: () => (
          <FormControl disabled>
            <Checkbox id="chk1" label="Check 1" />
            <Checkbox id="chk2" label="Check 2" />
          </FormControl>
        ),
      },
    ],
  },
  snippets: [
    {
      label: "Set error state of children",
      Code: () => (
        <FormControl error>
          <Checkbox id="chk1" label="Check 1" helperText="some description" />
          <Checkbox id="chk2" label="Check 2" />
        </FormControl>
      ),
    },
  ],
};
