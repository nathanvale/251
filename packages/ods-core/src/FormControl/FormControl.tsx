import * as React from "react";
import {
  FormControl as MuiFormControl,
  FormControlProps as MuiFormControlProps,
} from "@material-ui/core";
import {
  BaseFormStateProps,
  FormGroupElements,
} from "@origin-digital/ods-types";

export interface FormControlProps extends BaseFormStateProps {
  component?: FormGroupElements;
  muiProps?: MuiFormControlProps;
}

export const FormControl = ({
  muiProps = {} as any,
  ...others
}: FormControlProps) => <MuiFormControl {...muiProps} {...others} />;

FormControl.displayName = "FormControl";
FormControl.defaultProps = {
  component: "fieldset",
  "data-id": "form-control",
};
