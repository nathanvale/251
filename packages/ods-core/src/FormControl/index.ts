export { FormControl } from "./FormControl";
export type FormControlProps = import("./FormControl").FormControlProps;
