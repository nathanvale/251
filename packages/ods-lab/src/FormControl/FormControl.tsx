import * as React from "react";
import MuiFormControl, {
  FormControlProps as IFormControl,
} from "@material-ui/core/FormControl";

export interface FormControlProps extends IFormControl {
  "data-id"?: string;
  component?: React.ElementType;
}

export const FormControl = (props: FormControlProps) => (
  <MuiFormControl {...props}>{props.children}</MuiFormControl>
);
