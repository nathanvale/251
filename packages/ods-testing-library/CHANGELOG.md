# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.3.0](https://bitbucket.orgn.io/od/origin-ui/compare/@origin-digital/ods-testing-library@0.2.0...@origin-digital/ods-testing-library@0.3.0) (2020-02-12)


### Features

* **tests, columns:** alignY default changed to top ([c23c964](https://bitbucket.orgn.io/od/origin-ui/commits/c23c964))





# 0.2.0 (2020-01-31)


### Features

* set up jest for snapshots ([6e6fbfa](https://bitbucket.orgn.io/od/origin-ui/commits/6e6fbfa))
