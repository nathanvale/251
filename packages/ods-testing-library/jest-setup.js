import { configure } from "@testing-library/react";

require("jest-styled-components");

configure({ testIdAttribute: "data-id" });
