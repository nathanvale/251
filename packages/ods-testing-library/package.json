{
  "name": "@origin-digital/ods-testing-library",
  "description": "ODS testing tools that enhance react-testing-library",
  "license": "UNLICENSED",
  "author": "Nathan Vale <nathan.vale@origin.com.au>",
  "version": "0.5.0",
  "main": "dist/cjs/index.js",
  "module": "dist/es/index.js",
  "files": [
    "dist",
    "jest-setup.js"
  ],
  "scripts": {
    "build": "yarn build:cjs && yarn build:es",
    "build:cjs": "BUILD_FORMAT=cjs origin-scripts build --monorepo --delete-dir-on-start --out-dir dist/cjs",
    "build:es": "BUILD_WEBPACK=true origin-scripts build --monorepo --out-dir dist/es"
  },
  "sideEffects": false,
  "typings": "dist/typings/index.d.ts",
  "dependencies": {
    "@babel/runtime-corejs3": "^7.8.3",
    "@testing-library/jest-dom": "^5.1.1"
  },
  "peerDependencies": {
    "@origin-digital/ods-themes": "*",
    "@testing-library/react": "9.x",
    "jest-styled-components": "6.x",
    "react": "^16.9.0"
  },
  "devDependencies": {
    "@testing-library/react": "9.x"
  },
  "publishConfig": {
    "registry": "https://artifactory.origindigital-dac.com.au/artifactory/api/npm/npm-internal/"
  }
}
