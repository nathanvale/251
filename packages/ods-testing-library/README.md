# @origin-digital/ods-testing-library

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

- [Installation](#installation)
- [Why do I need this package?](#why-do-i-need-this-package)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

This package enhances the render function of `react-testing-library` so that the `<OriginThemeProvider/>` is wrapped around your rendered component.

All other API features of `react-testing-library` are also included.

## Installation

Install the package in your project directory with:

```sh
yarn add @origin-digital/ods-testing-library
```

This package also has a peer dependency on the following:

```sh
// with yarn
yarn add @origin-digital/ods-core
yarn add -D @testing-library/react
yarn add -D jest-styled-components
```

## Why do I need this package?

**The problem**

When using react-testing-library and ODS with the jest testing framework it is necessary to wrap **all** your components in our `<OriginThemeProvider/>` - otherwise the react component cannot be rendered and tested.

**The solution**

`@origin-digital/ods-testing-library` creates a wrapper around the `react-testing-library` render function. It is here that we wrap the `<OriginThemeProvider/>`. All other API features of `react-testing-library` are also included.

This practice is recommended by the `react-testing-library` documentation itself...

https://testing-library.com/docs/native-testing-library/setup#custom-render

The origin-ui dogfoods this package itself but it can also be used in your own app to test by adding to your jest config:

```
{
...,
setupFilesAfterEnv: [@origin-digital/ods-testing-library/jest-setup],
...
}
```

and usage in your tests:

```
import React from 'react';
import {render} from '@origin-digital/ods-testing-library';
import {Box} from '../Box/Box';
import {Stack} from './Stack';

test('It can stack', () => {
  const {container} = render(
    <Stack>
      <Box backgroundColor="red" padding="medium" />
      <Box backgroundColor="orange" padding="medium" />
      <Box backgroundColor="blue" padding="medium" />
    </Stack>,
  );
  expect(container.firstChild).toMatchSnapshot();
});
```
