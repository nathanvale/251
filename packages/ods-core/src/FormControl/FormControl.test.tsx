import React from "react";
import { render } from "@origin-digital/ods-testing-library";
import { FormControl } from "./FormControl";

test("renders a fieldset with default data-id", () => {
  const { container } = render(<FormControl id="myfc">Some label</FormControl>);

  expect(container).toMatchSnapshot();
});

test("renders a div element", () => {
  const { container } = render(
    <FormControl id="mylbl" component="div">
      Some label
    </FormControl>,
  );

  expect(container.getElementsByTagName("fieldset")).toHaveLength(0);
  expect(container.getElementsByTagName("div")).toHaveLength(1);
});
