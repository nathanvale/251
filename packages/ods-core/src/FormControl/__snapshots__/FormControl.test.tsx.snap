// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a fieldset with default data-id 1`] = `
<div>
  <fieldset
    class="MuiFormControl-root"
    data-id="formControl"
    id="myfc"
  >
    Some label
  </fieldset>
</div>
`;
